brainiak/hyperparamopt/hpo.py

```diff
@@ -0,0 +1,349 @@
+    EI as the ratio of the likelihoods of the 2 distributions. We pick the
+    point with the best EI among the samples that is also not very close to
+    a point we have sampled earlier.
+
+    Arguments
+    ---------
+
+    x : 1D array
+      Samples generated from the distribution so far
+
+    y : 1D array
+      Loss values at the corresponding samples
+
+    min_limit : double, default : -inf
+      Minimum limit for the distribution
+
+    max_limit : double, default : +inf
+      Maximum limit for the distribution
+
+    Returns
+    -------
+
+    x_next : double
+      Next value to use for HPO
+
+    """
+
+    z = np.array(list(zip(x, y)), dtype=np.dtype([('x', float), ('y', float)]))
+    z = np.sort(z, order='y')
+    n = y.shape[0]
+    g = int(np.round(np.ceil(0.15 * n)))
+    ldata = z[0:g]
+    gdata = z[g:n]
+    lymin = ldata['y'].min()
+    lymax = ldata['y'].max()
+    weights = (lymax - ldata['y']) / (lymax - lymin)
+    lx = gmm_1d_distribution(ldata['x'], min_limit=min_limit,
+                             max_limit=max_limit, weights=weights)
+    gx = gmm_1d_distribution(gdata['x'], min_limit=min_limit,
+                             max_limit=max_limit)
+
+    samples = lx.get_samples(n=1000)
+    ei = lx(samples) / gx(samples)
+
+    h = (x.max() - x.min()) / (10 * x.size)
+    # assumes prior of x is uniform -- should change for different priors
+    # d = np.abs(x - samples[ei.argmax()]).min()
+    # CDF(x+d/2) - CDF(x-d/2) < 1/(10*x.size) then reject else accept
+    s = 0
+    while (np.abs(x - samples[ei.argmax()]).min() < h):
+        ei[ei.argmax()] = 0
+        s = s + 1
+        if (s == samples.size):
+            break
+    xnext = samples[ei.argmax()]
+
+    return xnext
+
+
+def fmin(loss_fn,
+         space,
+         max_evals,
+         trials,
+         init_random_evals=30,
+         explore_prob=0.2):
+    """Find the minimum of function through hyper paramter optimization
+
+    Arguments
+    ---------
+
+    loss_fn : function that takes in a dictionary and returns a real value
+             Function to be minimized
+
+    space : Dictionary specifying the range and distribution of
+            the hyperparamters
+
+    max_evals : int
+               Maximum number of evaluations of loss_fn allowed
+
+    trials : list
+             Holds the output of the optimization trials
+             Need not be empty to begin with, new trials are appended
+             at the end
+
+    init_random_evals : int, default 30
+                        Number of random trials to initialize the
+                        optimization
+
+    explore_prob : double in [0, 1], default 0.2
+                   Controls the exploration-vs-exploitation ratio
+                   Currently 20% of trails are random samples
+
+    Returns
+    -------
+
+    best : trial entry (dictionary of hyperparameters)
+           Best hyperparameter setting found
```

Comment:
Use a 4-space indent for the second and following lines of descriptions of parameters, returns, etc.